src/bit_vec.rs

```diff
@@ -68,72 +30,75 @@ impl<O: BitOrder, T: BitStore + ToByteSlice> Encode for BitVec<O, T> {
 		);
 		Compact(len as u32).encode_to(dest);
 
-		let byte_slice: &[u8] = self.as_slice().as_byte_slice();
+		// NOTE: doc of `BitSlice::as_slice`:
+		// > The returned slice handle views all elements touched by self
+		//
+		// Thus we are sure the slice doesn't contain unused elements at the end.
+		let slice = self.as_slice();
 
-		if cfg!(target_endian = "big") && mem::size_of::<T>() > 1 {
-			let mut vec_u8: Vec<u8> = byte_slice.into();
-			reverse_endian(&mut vec_u8[..], mem::size_of::<T>());
-			dest.write(&vec_u8);
-		} else {
-			dest.write(byte_slice);
-		}
+		encode_slice_no_len(slice, dest)
 	}
 }
 
-impl<O: BitOrder, T: BitStore + ToByteSlice> EncodeLike for BitVec<O, T> {}
-
-/// # WARNING
-///
-/// In bitvec v0.17.4 the only implementations of BitStore are u8, u16, u32, u64, and usize.
-/// This implementation actually only support u8, u16, u32 and u64, as encoding of usize
-/// is inconsistent between platforms.
-impl<O: BitOrder, T: BitStore + FromByteSlice> Decode for BitVec<O, T> {
-	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
-		<Compact<u32>>::decode(input).and_then(move |Compact(bits)| {
-			let bits = bits as usize;
-			let required_bytes = required_bytes::<T>(bits);
-
-			let mut vec_u8 = read_vec_from_u8s::<I, u8>(input, required_bytes)?;
+impl<O: BitOrder, T: BitStore + Encode> Encode for BitVec<O, T> {
+	fn encode_to<W: Output>(&self, dest: &mut W) {
+		self.as_bitslice().encode_to(dest)
+	}
+}
 
-			if cfg!(target_endian = "big") && mem::size_of::<T>() > 1 {
-				reverse_endian(&mut vec_u8[..], mem::size_of::<T>());
-			}
+impl<O: BitOrder, T: BitStore + Encode> EncodeLike for BitVec<O, T> {}
 
-			let mut aligned_vec: Vec<T> = vec![0u8.into(); required_bytes / mem::size_of::<T>()];
+/// Equivalent of `BitStore::MAX_BITS` on 32bit machine.
+const ARCH32BIT_BITSLICE_MAX_BITS: usize = 0x1fff_ffff;
 
-			unsafe {
-				let aligned_u8_ptr = aligned_vec.as_mut_ptr() as *mut u8;
-				for (i, v) in vec_u8.iter().enumerate() {
-					*aligned_u8_ptr.add(i) = *v;
-				}
+impl<O: BitOrder, T: BitStore + Decode> Decode for BitVec<O, T> {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		<Compact<u32>>::decode(input).and_then(move |Compact(bits)| {
+			// Otherwise it is impossible to store it on 32bit machine.
+			if bits as usize > ARCH32BIT_BITSLICE_MAX_BITS {
+				return Err("Attempt to decode a bitvec with too many bits".into());
 			}
-
-			let mut result = Self::from_vec(aligned_vec);
-			assert!(bits <= result.len());
-			unsafe { result.set_len(bits); }
+			let required_elements = required_elements::<T>(bits)? as usize;
+			let vec = decode_vec_with_len(input, required_elements)?;
+
+			let mut result = Self::try_from_vec(vec)
+				.map_err(|_| {
+					Error::from("UNEXPECTED ERROR: `bits` is less than
+					`ARCH32BIT_BITSLICE_MAX_BITS`; So BitVec must be able to handle the number of
+					segment needed for `bits` to be represented; qed")
+				})?;
```

Comment:
I'm not sure how to phrase it.
Basically we have ensure `bits` is less than `ARCH32BIT_BITSLICE_MAX_BITS`.
So the number of segment needed for `bits` should be less or equal to the number of segment needed for `ARCH32BIT_BITSLICE_MAX_BITS`. which must be less or equal to the maximum number of segment allowed.

Reply:
but I wanted to insist that it is an internal error. Maybe it is actually better to panic with `expect("...; qed")`

Reply:
As long as the user understood it is a limitation from the bitvec crate, it is fine.